### PR TITLE
Pytorch Engine reduce TTFT

### PR DESCRIPTION
**Pull request by grimoire**
prefill enough requests before decoding.

### CODE REVIEW

File: lmdeploy/pytorch/engine/engine.py [hunk 1]
```diff
@@ -636,7 +631,11 @@ async def __long_context_single_forward(inputs):
                 ret['logits'] = ret['logits'][:, last_token_loc]
             return ret
         else:
-            return await __long_context_single_forward(inputs)
+            ret = await __long_context_single_forward(inputs)
+            if not return_logits and not inputs.is_decoding:
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/engine.py:635]: what benefits can we get from this change?

File: lmdeploy/pytorch/engine/engine.py [hunk 2]
```diff
@@ -636,7 +631,11 @@ async def __long_context_single_forward(inputs):
...
+                last_token_loc = [-1]
```
Comment (AllentDan) [inline lmdeploy/pytorch/engine/engine.py:636]: Does it influence the following computation?

File: lmdeploy/pytorch/engine/engine.py [hunk 3]
```diff
@@ -828,16 +827,15 @@ def __need_logits(seqs: SeqList):
...
+                swap_out_map = scheduler_output.swap_out_map
                 prefill_interval = self.scheduler_config.prefill_interval
-                schedule_output = self.scheduler.schedule(
-                    is_prefill=is_prefill, prealloc_size=prefill_interval)
-                running: SeqList = schedule_output.running
-                adapters = schedule_output.adapters
                 loop_count = 1 if is_prefill else (prefill_interval - 1)
-                if len(running) == 0:
-                    raise NoRunningSeqs()
+                assert len(running) > 0
```
Comment (AllentDan) [inline lmdeploy/pytorch/engine/engine.py:838]: how do we make sure it is True?

File: lmdeploy/pytorch/engine/engine.py (hunk 1 above)
Comment (grimoire) [reply lmdeploy/pytorch/engine/engine.py:635]: https://github.com/InternLM/lmdeploy/blob/9cb25830bf61a86fbf4bf4a4de801ecea4a38d31/lmdeploy/pytorch/engine/engine.py#L518-L525

File: lmdeploy/pytorch/engine/engine.py (hunk 2 above)
Comment (grimoire) [reply lmdeploy/pytorch/engine/engine.py:636]: __long_context_single_forward is used for very long prefill. Only last output is required.

File: lmdeploy/pytorch/engine/engine.py (hunk 3 above)
Comment (grimoire) [reply lmdeploy/pytorch/engine/engine.py:838]: https://github.com/InternLM/lmdeploy/blob/9cb25830bf61a86fbf4bf4a4de801ecea4a38d31/lmdeploy/pytorch/engine/engine.py#L934-L936

Empty requests would be skipped.

**Review by AllentDan — APPROVED** [review]

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**